src/vcl/Image.cc

```diff
@@ -31,19 +31,261 @@
 
 #include "vcl/Image.h"
 #include "vcl/Exception.h"
-#include "ImageData.h"
 
 using namespace VCL;
 
+    /*  *********************** */
+    /*        OPERATION         */
+    /*  *********************** */
+
+    /*  *********************** */
+    /*       READ OPERATION     */
+    /*  *********************** */
+
+Image::Read::Read(const std::string& filename, Image::Format format)
+    : Operation(format),
+      _fullpath(filename)
+{
+}
+
+void Image::Read::operator()(Image *img)
+{
+    if ( _format == Image::Format::TDB ) {
+        if ( img->_tdb == NULL )
+            throw VCLException(TileDBNotFound, "Image::Format indicates image \
+                stored in TDB format, but no data was found");
+
+        img->_tdb->read();
+        img->_height = img->_tdb->get_image_height();
+        img->_width = img->_tdb->get_image_width();
+        img->_channels = img->_tdb->get_image_channels();
+    }
+    else {
+        cv::Mat img_read = cv::imread(_fullpath, cv::IMREAD_ANYCOLOR);
+        img->shallow_copy_cv(img_read);
+        if ( img->_cv_img.empty() )
+            throw VCLException(ObjectEmpty, _fullpath + " could not be read, \
+                object is empty");
+    }
+}
+
+    /*  *********************** */
+    /*       WRITE OPERATION    */
+    /*  *********************** */
+
+Image::Write::Write(const std::string& filename, Image::Format format,
+    Image::Format old_format, bool metadata)
+    : Operation(format),
+      _old_format(old_format),
+      _metadata(metadata),
+      _fullpath(filename)
+{
+}
+
+void Image::Write::operator()(Image *img)
+{
+    if (_format == Image::Format::TDB) {
+        if ( img->_tdb == NULL ) {
+            img->_tdb = new TDBImage(_fullpath);
+            img->_tdb->set_compression(img->_compress);
+        }
+
+        if ( img->_tdb->has_data() )
+            img->_tdb->write(_fullpath, _metadata);
+        else
+            img->_tdb->write(img->_cv_img, _metadata);
+    }
+    else {
+        cv::Mat cv_img;
+        if (_old_format == Image::Format::TDB)
+            cv_img = img->_tdb->get_cvmat();
+        else
+            cv_img = img->_cv_img;
```

Comment:
I am not re-reviewing all this code since we have done it in the past. I am guessing that hasn't changed beyond its move into the up-level class.

Reply:
Have you seen any performance improvement with this change btw? Just curious

Reply:
> I am not re-reviewing all this code since we have done it in the past. I am guessing that hasn't changed beyond its move into the up-level class.

I copy-pasted these after Luis' patch, so yes, they're not changed except getting moved.

Reply:
> Have you seen any performance improvement with this change btw? Just curious

Yes, a pretty minuscule amount though. Since I pushed this change almost a month ago, I can't remember the exact improvement. Please check the test logs I attached. They're far from proper benchmarks, but give a rough idea on the execution time change.